spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/TestSparkVariants.java

```diff
@@ -0,0 +1,201 @@
+    array.add(Variants.of("hello"));
+    array.add(Variants.of((byte) 42));
+    array.add(Variants.ofNull());
+
+    testVariantRoundTrip(metadata, array);
+  }
+
+  @Test
+  public void testVariantObjectRoundTrip() {
+    VariantMetadata metadata = Variants.metadata("name", "age", "active");
+    ShreddedObject object = Variants.object(metadata);
+    object.put("name", Variants.of("John Doe"));
+    object.put("age", Variants.of((byte) 30));
+    object.put("active", Variants.of(true));
+
+    testVariantRoundTrip(metadata, object);
+  }
+
+  @Test
+  public void testVariantNestedStructures() {
+    VariantMetadata metadata = Variants.metadata("user", "scores", "address", "city", "state");
+
+    // Create nested object: address
+    ShreddedObject address = Variants.object(metadata);
+    address.put("city", Variants.of("Anytown"));
+    address.put("state", Variants.of("CA"));
+
+    // Create array of scores
+    ValueArray scores = Variants.array();
+    scores.add(Variants.of((byte) 95));
+    scores.add(Variants.of((byte) 87));
+    scores.add(Variants.of((byte) 92));
+
+    // Create main object
+    ShreddedObject mainObject = Variants.object(metadata);
+    mainObject.put("user", Variants.of("Jane"));
+    mainObject.put("scores", scores);
+    mainObject.put("address", address);
+
+    testVariantRoundTrip(metadata, mainObject);
+  }
+
+  @ParameterizedTest
+  @FieldSource("UNSUPPORTED_PRIMITIVES")
+  public void testUnsupportedOperations(VariantPrimitive<?> primitive) {
+    // This tests the current state where Spark integration is not fully implemented
+    // TIME, nano timestamps are not supported in Spark
+    ByteBuffer valueBuffer =
+        ByteBuffer.allocate(primitive.sizeInBytes()).order(ByteOrder.LITTLE_ENDIAN);
+    primitive.writeTo(valueBuffer, 0);
+
+    org.apache.spark.types.variant.Variant sparkVariant =
+        new org.apache.spark.types.variant.Variant(
+            valueBuffer.array(), VariantTestUtil.emptyMetadata().array());
+
+    assertThatThrownBy(sparkVariant::getType)
+        .as("Unsupported variant type in Spark")
+        .isInstanceOf(SparkRuntimeException.class);
+  }
+
+  private void testVariantRoundTrip(VariantMetadata metadata, VariantValue value) {
+    // Create Iceberg variant
+    Variant icebergVariant = Variant.of(metadata, value);
+
+    // Serialize to bytes
+    ByteBuffer metadataBuffer =
+        ByteBuffer.allocate(metadata.sizeInBytes()).order(ByteOrder.LITTLE_ENDIAN);
+    metadata.writeTo(metadataBuffer, 0);
+
+    ByteBuffer valueBuffer =
+        ByteBuffer.allocate(value.sizeInBytes()).order(ByteOrder.LITTLE_ENDIAN);
+    value.writeTo(valueBuffer, 0);
+
+    // Create Spark VariantVal from the same bytes
+    byte[] metadataBytes = metadataBuffer.array();
+    byte[] valueBytes = valueBuffer.array();
+    VariantVal sparkVariant = new VariantVal(valueBytes, metadataBytes);
+
+    GenericsHelpers.assertEquals(icebergVariant, sparkVariant);
```

Comment:
The assertions can be moved into this test, I think.

Reply:
I'm comparing the content of Iceberg variant and Spark variant here.